plugins/volume/org.xen.xapi.storage.fsp/sr.py

```diff
@@ -60,8 +65,41 @@ def detach(self, dbg, sr):
         pass
 
     def ls(self, dbg, sr):
-        return COWVolume.ls(
-            dbg, sr, importlib.import_module('fsp').Callbacks())
+        # check for broken symlinks
+        self._check_paths(sr)
+        # create new VDIs
+        self._scan(dbg, sr)
+        cb = importlib.import_module('fsp').Callbacks()
+        results = []
+        with VolumeContext(cb, sr, 'r') as opq:
+            with cb.db_context(opq) as db:
+                vdis = db.get_all_vdis()
+                all_custom_keys = db.get_all_vdi_custom_keys()
+
+            for vdi in vdis:
+                _vdi_sanitize(vdi, opq, db, cb)
+
+                image_format = ImageFormat.get_format(vdi.image_type)
+
+                vdi_uri = cb.getVolumeUriPrefix(opq) + vdi.uuid
+                custom_keys = {}
+                if vdi.uuid in all_custom_keys:
+                    custom_keys = all_custom_keys[vdi.uuid]
+
+                results.append({
+                    'uuid': vdi.uuid,
+                    'key': vdi.uuid,
+                    'name': vdi.name,
+                    'description': vdi.description,
+                    'read_write': True,
+                    'virtual_size': vdi.volume.vsize,
+                    'physical_utilisation': 0,
+                    'uri': [image_format.uri_prefix + vdi_uri],
+                    'keys': custom_keys,
+                    'sharable': bool(vdi.sharable)
+                })
+
+        return results
```

Comment:
I moved the SR.ls() from the COWVolume to the SR driver due to psize. This is symmetric with SR.stat().

Reply:
You can remove all this code using this piece instead:
```python
def ls(self, dbg, sr):
    self._check_paths(sr)
    self._scan(dbg, sr)
    return COWVolume.ls(dbg, sr, fsp.Callbacks())
```
And with a `fsp.py` file:

```python
import xapi.storage.libs.libcow.callbacks

class Callbacks(xapi.storage.libs.libcow.callbacks.Callbacks):
    def getVolumeUriPrefix(self, opq):
        return "fsp/" + opq + "|"
    
    # With this definition, the default `ls` method can call `volumeGetPhysSize` without error.
    # Otherwise an exception is thrown because the path is a folder instead of VHD or raw device.
    def volumeGetPhysSize(self, opq, name):
        return 0
```